src/net/server/service.rs

```diff
@@ -0,0 +1,624 @@
+where
+    F: Fn(
+        Request<Message<RequestOctets>>,
+    ) -> Result<
+        Transaction<
+            Result<CallResult<RequestOctets, Target>, ServiceError>,
+            Future,
+        >,
+        ServiceError,
+    >,
+    RequestOctets: AsRef<[u8]>,
+    Target: Composer + Default + Send + Sync + 'static,
+    Future: std::future::Future<
+            Output = Result<CallResult<RequestOctets, Target>, ServiceError>,
+        > + Send,
+{
+    type Target = Target;
+    type Future = Future;
+
+    fn call(
+        &self,
+        message: Request<Message<RequestOctets>>,
+    ) -> Result<
+        Transaction<
+            Result<CallResult<RequestOctets, Self::Target>, ServiceError>,
+            Self::Future,
+        >,
+        ServiceError,
+    > {
+        (*self)(message)
+    }
+}
+
+//------------ ServiceError --------------------------------------------------
+
+/// An error reported by a [`Service`].
+#[derive(Debug)]
+pub enum ServiceError {
+    /// The service was unable to parse the request.
+    FormatError,
+
+    /// The service encountered a service-specific error condition.
+    InternalError,
+
+    /// The service was unable to assemble the response.
+    NotImplemented,
+
+    /// The service declined to handle the request.
+    Refused,
+}
+
+impl ServiceError {
+    pub fn rcode(&self) -> Rcode {
+        match self {
+            Self::FormatError => Rcode::FormErr,
+            Self::InternalError => Rcode::ServFail,
+            Self::NotImplemented => Rcode::NotImp,
+            Self::Refused => Rcode::Refused,
+        }
+    }
+}
+
+//--- Display
+
+impl Display for ServiceError {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        match self {
+            Self::FormatError => write!(f, "Format error"),
+            Self::InternalError => write!(f, "Internal error"),
+            Self::NotImplemented => write!(f, "Not implemented"),
+            Self::Refused => write!(f, "Refused"),
+        }
+    }
+}
+
+//--- From<PushError>
+
+impl From<PushError> for ServiceError {
+    fn from(_: PushError) -> Self {
+        Self::InternalError
+    }
+}
+
+//--- From<ParseError>
+
+impl From<ParseError> for ServiceError {
+    fn from(_: ParseError) -> Self {
+        Self::FormatError
+    }
+}
+
+//------------ ServerCommand ------------------------------------------------
+
+/// Command a server to do something.
+#[derive(Copy, Clone, Debug)]
+pub enum ServerCommand<T: Sized> {
+    #[doc(hidden)]
+    /// This command is for internal use only.
+    Init,
+
+    /// Command the server to alter its configuration.
+    Reconfigure(T),
+
+    /// Command the connection handler to terminate.
+    ///
+    /// This command is only for connection handlers for connection-oriented
+    /// transport protocols, it should be ignored by servers.
+    CloseConnection,
+
+    /// Command the server to terminate.
+    Shutdown,
```

Comment:
If I recall correctly this was created after initial design discussions many months ago with @partim.

There are two ways to shut down a server:

1. From the application code that owns the servers and services.
2. From the service code.

If the application is structured such that it spawns servers and a service and lets the service handle things like receiving CTRL-C or some HTTP REST command to re-initialize or shutdown or something, it may be more convenient for the command to come from the service down to the servers than for the application main code to send it up frmo the application to the servers.

I was actually wondering about the utility of this myself and discussed it recently with @partim who felt it had value and should be kept.

Reply:
Thinking about this now, I’m tempted to change my mind. A shutdown request needs to be processed by the listener rather than a connection, so maybe this should somehow be part of the way that the listener is being driven?

Maybe take it out here now and add a “clean shutdown” to the open issues? You can already do a hard shut down by dropping the listener, so this should be fine?

Reply:
I will remove it.

Note that the servers have a `fn shutdown()` that sends `ServerCommand::Shutdown` (not `ServiceFeedback::Shutdown` which is also received by in-progress connections. Connections will terminate after flushing any pending writes, and servers will stop accepting new connections and exit.

A`Drop` of a StreamServer does not currently cause the connections to exit but they will fail to monitor the ServerCommand channel because the server end will have dropped and will then disconnect with flush.

Reply:
Removed.